### ISSUE TITLE
issues.exclude-dirs not behaving as expected

### ISSUE THREAD

**Issue by gakesson** [golangci/golangci-lint#5438]
### Welcome

- [x] Yes, I'm using a binary release within 2 latest [releases](https://github.com/golangci/golangci-lint/releases). Only such installations are supported.
- [x] Yes, I've searched similar [issues on GitHub](https://github.com/golangci/golangci-lint/issues) and didn't find any.
- [x] Yes, I've read the `typecheck` section of the [FAQ](https://golangci-lint.run/welcome/faq/#why-do-you-have-typecheck-errors).
- [x] Yes, I've tried with the standalone [linter](https://golangci-lint.run/usage/linters/) if available (e.g., gocritic, go vet, etc.).
- [x] I agree to follow this project's [Code of Conduct](https://github.com/golangci/golangci-lint?tab=coc-ov-file)

### How did you install golangci-lint?

Official binary

### Description of the problem

The issues.exclude-dirs configuration is ignored in case lint failure is due to typecheck linter. 
This is a bit unintuitive that even explicitly excluded directories are failing a specific linter, and after some digging in the golangci-lint code it seems it is related to [this PR](https://github.com/golangci/golangci-lint/pull/4870).

This is a bit problematic when there is CI/CD tasks populate code that does not pass this linter, in build/CI/CD related directories not intended to be linted. Only workaround I have right now is to explicitly point out the directories that should be linted, which is a bit tedious.

Is this considered a bug or is it possible to have another config-option to instruct golangci-lint to not use typecheck linter even in certain directories?

### Version of golangci-lint

<details>

```console
$ golangci-lint --version
golangci-lint has version 1.64.4 built with go1.24.0 from 04aec4f7 on 2025-02-12T21:58:21Z
```

</details>

### Configuration

<details>

```console
run:
  tests: false
  timeout: 5m
issues:
  exclude-dirs:
    - build

```

</details>

### Go environment

<details>

```console
$ go version && go env
go version && go env
go version go1.23.6 linux/amd64
GO111MODULE=''
GOARCH='amd64'
GOBIN=''
GOCACHE='/home/user/.cache/go-build'
GOENV='/home/user/.config/go/env'
GOEXE=''
GOEXPERIMENT=''
GOFLAGS=''
GOHOSTARCH='amd64'
GOHOSTOS='linux'
GOINSECURE=''
GOMODCACHE='/home/user/golang/pkg/mod'
GONOPROXY='comp.com/*'
GONOSUMDB='comp.com/*'
GOOS='linux'
GOPATH='/home/user/golang'
GOPRIVATE='comp.com/*'
GOPROXY='https://proxy.golang.org,direct'
GOROOT='/home/user/golang/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.6.linux-amd64'
GOSUMDB='sum.golang.org'
GOTMPDIR=''
GOTOOLCHAIN='auto'
GOTOOLDIR='/home/user/golang/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.6.linux-amd64/pkg/tool/linux_amd64'
GOVCS=''
GOVERSION='go1.23.6'
GODEBUG=''
GOTELEMETRY='local'
GOTELEMETRYDIR='/home/user/.config/go/telemetry'
GCCGO='gccgo'
GOAMD64='v1'
AR='ar'
CC='gcc'
CXX='g++'
CGO_ENABLED='1'
GOMOD='/home/user/git/org/bridge/go.mod'
GOWORK=''
CGO_CFLAGS='-O2 -g'
CGO_CPPFLAGS=''
CGO_CXXFLAGS='-O2 -g'
CGO_FFLAGS='-O2 -g'
CGO_LDFLAGS='-O2 -g'
PKG_CONFIG='pkg-config'
GOGCCFLAGS='-fPIC -m64 -pthread -Wl,--no-gc-sections -fmessage-length=0 -ffile-prefix-map=/tmp/go-build3213195946=/tmp/go-build -gno-record-gcc-switches
```

</details>

### Verbose output of running

<details>

```console
$ golangci-lint cache clean
$ golangci-lint run -v
N/A
```

</details>

### A minimal reproducible example or link to a public repository

<details>

```go
N/A - see issue description
```

</details>

### Validation

- [x] Yes, I've included all information above (version, config, etc.).

### Supporter

- [ ] I am a sponsor/backer through [GitHub](https://github.com/sponsors/golangci/) or [OpenCollective](https://opencollective.com/golangci-lint)

**Comment by bombsimon** [comment]
You've checked the "Yes, I've read the typecheck section of the [FAQ](https://golangci-lint.run/welcome/faq/#why-do-you-have-typecheck-errors)" box but did you really do that? If so it seems like you didn't understand the documentation.

The key point is that [`typecheck` is not a real linter](https://github.com/golangci/golangci-lint/issues?q=is%3Aissue+%22typecheck+is+not+a+real+linter). You will never be able to disable it or exclude it since it's not a linter.

> ...  in build/CI/CD related directories not intended to be linted

Are these directories also not supposed to be able to compile with Go? If not, you will have to invoke `golangci-lint` and target specific packages, e.g. `golangci-lint run ./pkg1 ./pkg2`. Default is `./...` which will include all packages package, including ones that doesn't build successfully.

If the code _is_ supposed to be valid and compile successfully, can you confirm that `go build ./...` works and that you've followed the checklist in the FAQ on how to troubleshoot `typecheck` errors?

**Comment by bombsimon** [comment]
Gotcha!

To clarify the reason why this is happening is that the `exclude-dir` setting is referring to excluding _issues_ from these directories, not the actual analysis, note that it's under the `issues` section.

Running `golangci-lint` (or `golangci-lint ./...`) would result in a similar behavior as `go build ./...`. To avoid `golangci-lint` to analyze specific packages you would simply have to use a different pattern than `./...` just like when you target `./cmd` to build.

Best I can think of from the top of my head would be something like

```sh
golangci-lint run $(find . -maxdepth 1 -type d ! -name "*some-exclude-pattern*" ! -name '.*' -exec echo "{}/..." \;)
```

Or maybe easier to write include patterns?

```sh
golangci-lint run $(find . -maxdepth 1 -type d -name "*some-include-pattern*" -exec echo "{}/..." \;)
```

**Comment by ldez** [comment]
You can also use build tags to skip the "non-compilable" files.
